File: sqlite-android-plugin/src/main/java/com.alecstrong.sqlite.android/lang/SqliteParserDefinition.java
package com.alecstrong.sqlite.android.lang;

import com.alecstrong.sqlite.android.SQLiteLexer;
import com.alecstrong.sqlite.android.SQLiteParser;
import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.extapi.psi.PsiElementBase;
import com.intellij.lang.*;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.antlr.intellij.adaptor.lexer.ElementTypeFactory;
import org.antlr.intellij.adaptor.lexer.SimpleAntlrAdapter;
import org.antlr.intellij.adaptor.lexer.TokenElementType;
import org.antlr.intellij.adaptor.parser.AntlrParser;
import org.antlr.intellij.adaptor.parser.SyntaxErrorListener;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;

public class SqliteParserDefinition implements ParserDefinition {
	public static final IFileElementType FILE =
			new IFileElementType(SqliteLanguage.INSTANCE);
	@NotNull
	@Override
	public Lexer createLexer(Project project) {
		return new SimpleAntlrAdapter(SqliteLanguage.INSTANCE, new SQLiteLexer(null));
	}

	@Override
	public PsiParser createParser(Project project) {
		return new SqliteParser();
	}

	@NotNull
	@Override
	public TokenSet getWhitespaceTokens() {
		return ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE, Arrays.asList(SQLiteLexer.tokenNames), SQLiteLexer.SPACES);
	}

	@NotNull
	@Override
	public TokenSet getCommentTokens() {
		return ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE, Arrays.asList(SQLiteLexer.tokenNames), SQLiteLexer.SINGLE_LINE_COMMENT, SQLiteLexer.MULTILINE_COMMENT);
	}

	@NotNull
	@Override
	public TokenSet getStringLiteralElements() {
		return ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE, Arrays.asList(SQLiteLexer.tokenNames), SQLiteLexer.SINGLE_LINE_COMMENT, SQLiteLexer.STRING_LITERAL);
	}

	@Override
	public IFileElementType getFileNodeType() {
		return FILE;
	}

	@Override
	public PsiFile createFile(FileViewProvider viewProvider) {
		return new SqliteFile(viewProvider);
	}

	@NotNull
	@Override
	public PsiElement createElement(ASTNode node) {
		return new ASTWrapperPsiElement(node);
	}

	@Override
	public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode left, ASTNode right) {
		return SpaceRequirements.MAY;
	}
}

File: sqlite-android-studio-plugin/src/main/kotlin/com/squareup/sqlite/android/lang/SqliteParserDefinition.kt
/*
 * Copyright (C) 2016 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.sqlite.android.lang

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.tree.IFileElementType
import com.squareup.sqlite.android.SQLiteLexer
import org.antlr.intellij.adaptor.lexer.ElementTypeFactory
import org.antlr.intellij.adaptor.lexer.SimpleAntlrAdapter

class SqliteParserDefinition : ParserDefinition {
  private val file = IFileElementType(SqliteLanguage.INSTANCE)

  override fun createLexer(project: Project) =
      SimpleAntlrAdapter(SqliteLanguage.INSTANCE, SQLiteLexer(null))

  override fun createParser(project: Project) = SqliteParser()

  override fun getWhitespaceTokens() = ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE,
      listOf(*SQLiteLexer.tokenNames), SQLiteLexer.SPACES)

  override fun getCommentTokens() = ElementTypeFactory.createTokenSet(SqliteLanguage.INSTANCE,
      listOf(*SQLiteLexer.tokenNames), SQLiteLexer.SINGLE_LINE_COMMENT,
      SQLiteLexer.MULTILINE_COMMENT)

  override fun getStringLiteralElements() = ElementTypeFactory.createTokenSet(
      SqliteLanguage.INSTANCE, listOf(*SQLiteLexer.tokenNames), SQLiteLexer.SINGLE_LINE_COMMENT,
      SQLiteLexer.STRING_LITERAL)

  override fun getFileNodeType() = file

  override fun createFile(viewProvider: FileViewProvider) = SqliteFile(viewProvider)

  override fun createElement(node: ASTNode) = node.asPSINode()

  override fun spaceExistanceTypeBetweenTokens(left: ASTNode, right: ASTNode) =
      ParserDefinition.SpaceRequirements.MAY
}
